### ISSUE TITLE
Parameter selection to Improve the segmentation

### ISSUE THREAD

**Issue by CheLamVien** [DanuserLab/u-segment3D#4]
Dear all, 

Thank you for the excellent Python package. I have tried to use [segment_epithelial_organoid_2D_stacks_with_cellpose_auto-diameter.py](https://github.com/DanuserLab/u-segment3D/blob/master/tutorials/anisotropic/segment_epithelial_organoid_2D_stacks_with_cellpose_auto-diameter.py) to segmented the osteoblast-like cells, which adhere to the bottom of a chamber. The segmentation of the first few slices of the z-stack is perfect. However, I would like to improve the results of the middle slice.  The mask of the cell on the right in the cluster of three cells would be a bit bigger. I have tried to increase the `cellpose_segment_params['ksize']` as well as change to different `cellpose_segment_params['cellpose_modelname'] `, but I could not succeed.
![Screenshot from 2024-09-25 16-21-33](https://github.com/user-attachments/assets/c2f4f437-3ecc-4d99-8b4e-e6d04e336b29)

I could improve the segmentation using Cellpose by re-training the model and then manually selecting different `diameters ` for different slices.
I have also tried my re-trained model in `u-segment3D` using `model = models.CellposeModel(gpu=cellpose_segment_params['gpu'], pretrained_model=modeldir)` but the results are even worse than using the pre-trained model.

Is there any way to improve the segmentation within `u-segment3D`?

Thank you so much for your support.

**Comment by fyz11** [comment]
Howdy, 

Thanks for reaching out! 
Here's a couple of thoughts based on your screenshot:

**Try setting one optimal diameter on a best 2D slice and use the direct method**
1) am i right in thinking you have a single layer of cells? such that your side profile (i.e. viewing the cell side on is of a hemispherical-ish profile? From your description it sounds like the auto diameter is being thrown off - presumably, the cell gets gradually smaller nearer the top, reducing the number of foreground pixels. If this is the side profile of these cells, i'd recommend trying [segment_unwrapped_drosophila_2D_stacks.py](https://github.com/DanuserLab/u-segment3D/blob/master/tutorials/anisotropic/segment_unwrapped_drosophila_2D_stacks.py), with cellpose_segment_params['use_edge'] = True. This uses edge strength to select a best 2D slice and sets one optimal diameter for the whole stack. Then perform the aggregation in xy only (assuming your other views are bad). 

**Try using the u-Segment3D contrast method to get optimal diameter. Your current script isn't actually using ksize at the moment, : )**
[segment_epithelial_organoid_2D_stacks_with_cellpose_auto-diameter.py](https://github.com/DanuserLab/u-segment3D/blob/master/tutorials/anisotropic/segment_epithelial_organoid_2D_stacks_with_cellpose_auto-diameter.py) uses Cellpose's regression model to infer diameter. I've found this erratic when there's not much foreground pixels. A more stable method is our auto-contrast method whose example is this one: [segment_epithelial_organoid_2D_stacks.py](https://github.com/DanuserLab/u-segment3D/blob/master/tutorials/anisotropic/segment_epithelial_organoid_2D_stacks.py). This is where ksize comes in 

**Try changing factor=0.5, to factor =0.25 in preprocessing step**
Cellpose is trained on resizing images to fixed diameter = 30 pixels. From your screenshot, it seems your cells are much larger, more like 120pixels diameter. downsizing the images will greatly help because of two reasons: i) it blurs out internal detail which throws off segmentation and ii) you segment large objects easier!. This often boost segmentation better than biasing ksize (more a local refinement, and you'd have to whack it up greatly like ksize=55-105 for these large cells). 

**In preprocessing, further boosting the contrast with gamma correction + adaptive histogram equalization**
Cellpose is sensitive to illumination uneveness and contrast of edges. You can see the center of your cells are brighter than the edges and have weaker edges. Therefore after the normalization step, I might suggest trying skimage.exposure.adjust_gamma(img_preprocess, gamma=0.5) to amplify the background pixels. You can also try skimage.exposure.equalize_adapthist(img_preprocess, clip_limit=0.02) to also boost local contrast. 
![image](https://github.com/user-attachments/assets/b8b5f82e-fb19-44af-90e2-0763553d44a8)

**Use a lower threshold when computing the foreground binary**
Related to the above, the thresholding of the foreground cell probability dictates the accuracy of your cell boundaries. It seems you have accurate delineation of individual cells, just lacking weaker intensities. This could be because in this region the cell probability predicted by cellpose is lower. The way to change this is:
![image](https://github.com/user-attachments/assets/241ad664-4bfa-4ac8-a313-bb66dbdf0c51)
In the example script, I bin probability into 3 classes and use the highest bin as threshold. If you set aggregation_params['combine_cell_probs']['threshold_level'] = 0, you can use the lower threshold and this may capture better those weaker regions. If this lowers the threshold too much, you can alternatively drop aggregation_params['combine_cell_probs']['threshold_n_levels'] = 3 to aggregation_params['combine_cell_probs']['threshold_n_levels'] = 2. 


**If you have already trained a 2D model and getting nice slice-to-slice segmentations, don't bother with the initial steps, go direct to the 2D-to-3D aggregation using the indirect method**
We provide cellpose as a convenience so users have a 2D segmentation tool. Its definitely the most finicky to tune in the entire pipeline : ). If you have a better 2D segmented stacks, please just aggregate on that to get your 3D segmentation :D i.e. in the script you show, go straight to line 323 and run whats below that, setting `labels_xy` to be your segmented 2D stack.  
![image](https://github.com/user-attachments/assets/d551abed-466a-43f1-a8fc-a78e9863e8b3). If you want to still play with auto-tuning diameter, i'd again recommend doing the u-Segment3D auto-contrast method in the other example script. 

Let me know if these work for you !

**Comment by CheLamVien** [comment]
Thank you so much for your detailed suggestion. 

You are right; I only have a single layer, and the view of the cell side is of a hemispherical profile.

I have tried [segment_unwrapped_drosophila_2D_stacks.py](https://github.com/DanuserLab/u-segment3D/blob/master/tutorials/anisotropic/segment_unwrapped_drosophila_2D_stacks.py). 

I used the preprocess steps from [segment_epithelial_organoid_2D_stacks_with_cellpose_auto-diameter.py](https://github.com/DanuserLab/u-segment3D/blob/master/tutorials/anisotropic/segment_epithelial_organoid_2D_stacks_with_cellpose_auto-diameter.py). In additions, `img_preprocess = skexposure.adjust_gamma(img_preprocess, gamma=0.5)` helps. I have also tried to apply image deconvolution using `unsupervised_wiener`. The results are getting better. 

    def wiener_3d(image_3d, psf, axis=0):
        restored_slices = []
        
        # Iterate over each 2D slice along the given axis
        for i in range(image_3d.shape[axis]):
            if axis == 0:
                slice_2d = image_3d[i, :, :]
            elif axis == 1:
                slice_2d = image_3d[:, i, :]
            elif axis == 2:
                slice_2d = image_3d[:, :, i]
            
            # Apply unsupervised Wiener filter on the 2D slice
            restored_slice, _ = unsupervised_wiener(slice_2d, psf)
            
            restored_slices.append(restored_slice)

        # Reconstruct the 3D volume by stacking the slices
        if axis == 0:
            return np.stack(restored_slices, axis=0)
        elif axis == 1:
            return np.stack(restored_slices, axis=1)
        elif axis == 2:
            return np.stack(restored_slices, axis=2)
    img_preprocess = uSegment3D_filters.normalize(img, clip=True)
    import skimage.exposure as skexposure

    img_preprocess = skexposure.adjust_gamma(img_preprocess, gamma=0.5)
    #img_preprocess = skexposure.equalize_adapthist(img_preprocess, clip_limit=0.02)
    factor = 0.5
    
    # the xy size is large - we resize, downsample each slice 2x  
    img_preprocess = np.array([ndimage.zoom(imm, zoom=[factor, factor], order=1, mode='reflect') for imm in img_preprocess])
    # the background is estimated with Gaussian with sigma 1/4 the size. You can also decrease or increase the fraction which might better correct but is slower computationally. 
    img_preprocess_bg_remove = np.array([np.nanmean(img_preprocess)/(ndimage.gaussian_filter(im_ch, sigma=im_ch.shape[1]//4) + 1)*im_ch for im_ch in img_preprocess]) # correct background. 
    img_preprocess_bg_remove = uSegment3D_filters.normalize(img_preprocess_bg_remove, clip=True)
    # use now the background corrected 
    img_preprocess = img_preprocess_bg_remove.copy()


    psf_size = 15
    psf = np.zeros((psf_size, psf_size)) 
    psf[psf_size//2, psf_size//2] = 1
    psf = ndimage.gaussian_filter(psf, sigma=1)**2.
    psf = psf/float(np.sum(psf))
    img_preprocess = wiener_3d(img_preprocess, psf, axis=0)


However, I encountered another problem: there are holes in the segmentation masks (even `aggregation_params['postprocess_binary']['binary_fill_holes'] = True`) and one cell was split. I have tried to increase `aggregation_params['gradient_descent']['gradient_decay'] = 0.05`, but then some of the cells were merged. :(

![image](https://github.com/user-attachments/assets/d9559e54-8e02-458a-aede-ba5a9200e940)

**If you have already trained a 2D model and are getting nice slice-to-slice segmentations, don't bother with the initial steps, go direct to the 2D-to-3D aggregation using the indirect method** I do want to find a way to determine the optimal diameter automatically as I have to manually set the diameter for each slice to get the good segmentation. I really want to avoid that. 

I attached the image that I want to segment ([link](https://drive.google.com/file/d/1cFadJCYgjAESWy9z-gPvTcPHqq32X8T1/view?usp=drive_link)). 

Thank you so much for your time and support!

**Comment by fyz11** [comment]
hm.... seeing your image, I see why you are having difficulty with cellpose tuning now. Cellpose is very sensitive to edges, and whats happening is its being thrown off by the brightness in the center. and the fact that in some slices, the cells are well separated but others they aren't. 

since i think now the main problem is diameter. I'm tempted to suggest, perform a slice-by-slice optimization to get out best_diam per slice. Plot this as a function of z-slice (i predict this is going to be noisy, but there might be a trend...). I would like smooth this e.g. moving average filter and use the smoothened values to compute actual segmentations ..... 

Could you also attach the script you used to get the result you showed? thanks! - so i can reproduce the actual segmentation you would be seeing :D

aggregation_params['gradient_descent']['gradient_decay'] has a large effect because its exponential.... . I typically start with 0 first as default, then try 0.01 (this will already have an effect), then 0.05, 0.1, 0.25, 0.5, 0.75,1 (something like this...)  : )

**Comment by fyz11** [comment]
btw - if you have a nuclear channel with these images, this could really help to improve the segmentation. I typically find two-color cellpose to work better.

**Comment by CheLamVien** [comment]
The images have only one channel.

After playing around with the parameters, I'm quite happy with the results. I have attached the script that I used. I would be more than happy if you could give me some feedback on the improvement of the segmentation (e.g., connecting two parts of the red cell). [segment_direct_2D_stacks.txt](https://github.com/user-attachments/files/17162913/segment_direct_2D_stacks.txt)

Regarding your suggestion to obtain the best_diam for each slice by performing slice-by-slice optimization and then using the smoothed values for actual segmentation, how can I implement this? Could you recommend a tutorial or guide that would help me get started? I'd like to give it a try as well.

Btw I found that post `process_segment_params['size_filters']['minsize']` does not filter cell 2D but 3D if one uses it in `uSegment3D.postprocess_3D_cell_segmentation`. 
![image](https://github.com/user-attachments/assets/8ab9885c-2e5b-483e-bf2a-64e90bf41640)
I have defined a function to do it, and I hope it helps 

```
def filter_small_objects_2d(label_image_3d, min_size):
    # Get the shape of the 3D image
    z, y, x = label_image_3d.shape
    filtered_label_image_3d = np.copy(label_image_3d)  # Copy the original image

    # Loop through each 2D slice
    for i in range(z):
        slice_2d = label_image_3d[i, :, :]  # Extract 2D slice
        
        # Label connected components in the 2D slice
        labeled_slice, num_features = label(slice_2d)
        
        props = regionprops(labeled_slice)
        
        for prop in props:
            if prop.area < min_size:
                # Set the small objects to 0 in the original slice
                filtered_label_image_3d[i, labeled_slice == prop.label] = 0
    return filtered_label_image_3d
```

**Comment by fyz11** [comment]
ok, i think i got the segmentation you were going after.... :D now. Please see the script in the .zip (as i cannot directly attach .py files)
[segment_test_3D_slice-by-slice.zip](https://github.com/user-attachments/files/17162932/segment_test_3D_slice-by-slice.zip)

I dropped the segmentation n_threshold  from 3 to 2 in the slice-by-slice to get more of the weaker signal. I also restricted the scan diameter range when auto-determining the optimal cellpose diameter in the u-Segment3D contrast-score method (to constrain the flexibility). There are still intermediate slices where 1 cell breaks into 2. We overcome this in the 2D-to-3D indirect method aggregation by simply using less gradient descent iterations (erode less) and adding some smoothing of xy gradients (encourage bleed through across z-slices).

My 3D visualization is: 
![image](https://github.com/user-attachments/assets/146e12e6-ff98-43a6-acc9-1b0f90ea8162)
![image](https://github.com/user-attachments/assets/82522b31-63bf-4393-a42c-01b443eb3da9)

The end result is probably larger than what you want. Theres two ways to handle this. 
1) separately derive a binary foreground by intensity thresholding the image. This is what i did in https://github.com/DanuserLab/u-segment3D/blob/master/tutorials/anisotropic/segment_single-cell_tracking_challenge_2D_stacks.py and then use this to mask the segmentation. 

2) Check out the postprocessing capability of u-Segment3D illustrated in https://github.com/DanuserLab/u-segment3D/blob/master/tutorials/single_cells/segment_blebs_3D.py. There are two steps, one is the label diffusion which can bring boundaries closer to edges in a 'guide image', then we can apply guide filter to 'snap' the segmentation.

**Comment by fyz11** [comment]
Sorry, I hadn't seen your previous post, before my last one. Hopefully my script answers your question of linking broken cells across z-slices in the 3D aggregation. Trick is smoothing the gradient and using less iterations. (essentially iterations just need to be enough to spatially separate actual neighbors). 

r.e. cell diameter sure. In the code I gave you, there is the following:
![image](https://github.com/user-attachments/assets/0ba86895-3e58-4b3f-a404-709815b63502)

Plotting it, the blue is the auto-inferred. The orange is the smoothened version, which shows an increase in diameter gradually. It looks to me by eye, theres 3? plateaus. You could apply otsu on this or kmeans clustering to identify 3 thresholds and the z-slices they apply to. In the script i gave you, I didn't need to do this, since the segmentations were already good enough for assembling into 3D. 
![image](https://github.com/user-attachments/assets/02e351bf-219e-4e33-8ca2-365e9bb550b9)

**Comment by CheLamVien** [comment]
Thank you so much for your support. The segmentation is good now. 




> You could apply otsu on this or kmeans clustering to identify 3 thresholds and the z-slices they apply to.

Could you explain how I can implement this approach in my script (in case I need it in the future)? I'm quite confused. 

> 1. separately derive a binary foreground by intensity thresholding the image. This is what i did in https://github.com/DanuserLab/u-segment3D/blob/master/tutorials/anisotropic/segment_single-cell_tracking_challenge_2D_stacks.py and then use this to mask the segmentation.

You mean this: 

    import skimage.morphology as skmorph 
    import skimage.color as skcolor 
    import seaborn as sns 
    
    # slight smoothing. 
    img_preprocess_smooth = ndimage.gaussian_filter(img_preprocess, sigma=1)
    mask3D = img_preprocess_smooth >= skfilters.threshold_multiotsu(img_preprocess_smooth, 3)[-1]
    mask3D = skmorph.binary_closing(mask3D, skmorph.ball(3))
    
    final_mask = mask3D*segmentation3D_filt
    final_mask = uSegment3D_filters.largest_component_vol_labels_fast(final_mask) # ensure each label is a single-component
    
    
    final_mask_color = np.uint8(255*skcolor.label2rgb(final_mask, 
                                                  colors=sns.color_palette('Spectral', n_colors=16),
                                                  bg_label=0))
    
    plt.figure(figsize=(10,10))
    plt.imshow(final_mask_color[20])
    plt.show()
        
    
    uSegment3D_fio.save_segmentation(os.path.join(savefolder_seg,
                                basename+'_binary_3D_Otsu_Foreground_mask.tif'),
                                     np.uint16(mask3D))
    
    # save the final segmentation, masked with binary 
    uSegment3D_fio.save_segmentation(os.path.join(savefolder_seg,
                            basename+'_cp_labels_%s_3D_consensus_with_Foreground_mask.tif' %(str(view))),
                                 np.uint16(final_mask))

**Comment by fyz11** [comment]
1. using smoothed diameter, or k diameters across z-slices.
- all_params_diams records the diameter parameter used in cellpose for each z-slice. If you have 50 z-slices, you have 50 diameters --- this is the blue line. You can see its very bumpy.....which is unrealistic (as we expect the cell size to change gradually).
- to enforce smoothness, we apply moving average with winsize=15 (for example). The winsize in this case is just smoothing the diameters across 15 slices on a sliding window basis. ---> this gives the orange curve.

You now have two options:
a) use the orange curve i.e. those smoothened diameters, and re-run cellpose for z-slice 1 with first diameter etc. 
so you would specify a value for `diameter` in model.eval and not leave it as None. 

b) use the orange curve to help determine a few diameters (this is like when you said, you chose 2/3 diameters for different z-slices. From the curve, it looks like we should choose 3 thresholds:

![image](https://github.com/user-attachments/assets/7505b656-df98-4769-be40-164524d5fa44)

We can find these thresholds by partitioning the diameters into groups/clusters. Multi-otsu c.f. https://scikit-image.org/docs/stable/auto_examples/segmentation/plot_multiotsu.html does exactly this:

thresholds = skimage.filters.threshold_mulitotsu(ma_of_diam, 3)    % this will give two thresholds

group1 = ma_of_diam[ma_of_diam<thresholds[0]
group2 = ma_of_diam[np.logical_and(ma_of_diam>thresholds[0], ma_of_diam<thresholds[1])]
group3 = ma_of_diam[ma_of_diam>thresholds[1]]

then use the mean of group1, group2, group3 as your diameter 1, 2, 3 respectively. Then for the corresponding z-slices you would set model.eval diameter argument as those.

2. Yes, that script is the one i meant for external segmentation. I would use it after you done your skexposure.adjust_gamma. You should be able to get a nice binary of where all cells are, and use it as  `mask3D` to multiply your other segmentation. This is because I find that cellpose doesn't generate very tight boundary masks when the input data is not alike to its training data, but thats fine here --- the main objective was to use it to tell where individual cells are within the clump. You can compute the foreground binary with any other method and mask.

**Comment by CheLamVien** [comment]
Thank you so much for your support and guidance. I'll go ahead and close this ticket now. I appreciate your assistance!
